### PR TITLE
feat: add Zodiac execution helpers

### PR DESCRIPTION
**Pull request by Sekhmet**
## Summary

This PR adds two new clients:
- `SpaceManager` for managing space on Starknet (currently implementing only finalizeProposal)
- `Zodiac` for doing L1 execution.

Closes https://github.com/snapshot-labs/sx.js/issues/94

## Test plan

- Use https://github.com/snapshot-labs/sx-ui/pull/348 to start proposal and vote til it reaches quorum.
- Modify `test/integration/spaceManager.test.ts` to match your proposal.
- Run `STARKNET_ADDRESS=xxx STARKNET_PK=xxx yarn test:integration test/integration/spaceManager.test.ts`
- Wait for it to be accepted on L1.
- Modify `test/integration/zodiac.test.ts` by adding `only` once for receiving proposal, and second time for executing specific transaction - you will need to modify space/proposalIndex/transaction to match - **proposalIndex use https://goerli.etherscan.io/address/0x196f0548e3140d2c7f6532a206dd54fbc12232a4#readContract to read it before calling receiveProposal**
- Run `ETH_PK=xxx GOERLI_NODE_URL=xxx yarn test:integration test/integration/zodiac.test.ts`

## Further comments

Sample execution:
 - `finalizeProposal`: https://testnet-2.starkscan.co/tx/0x02d81cd93004412e3ab9f76c0eb5da3fb144caebdd9136e96a2a4e4a677ea2d6#overview
 - `receiveProposal`: https://goerli.etherscan.io/tx/0x666717e31bfc6e6d4c987ce6190f2e59e9419d5cba94ee31838fdb478af7accc
 - `executeProposalTx`: https://goerli.etherscan.io/tx/0x0d3e0bae9bc6a2cd27fb5ca0ff26a74ed432e1b9985d2df678c48a0b4f04bdeb

### CODE REVIEW

File: src/clients/zodiac/abis/zodiac.json [hunk 1]
```diff
@@ -0,0 +1,389 @@
+[
+  {
```
Comment (bonustrack) [inline src/clients/zodiac/abis/zodiac.json:2]: Can we use string format here? It's much more compact, here is a function to turn this to string format  https://github.com/snapshot-labs/snapshot.js/blob/master/test/abi.ts

**Comment by bonustrack** [comment]
@Sekhmet I got this contract error for some reason when trying to finalize a proposal
image
I'm not sure what `2017523155088874421268775296176650446770562571530955066323122001281112303739` or `2010488055149106900438845871269516371322030882897905069258586578468989457483` refer to. The proposal is here: https://testnet.snapshotx.xyz/#/0x07e6e9047eb910f84f7e3b86cea7b1d7779c109c970a39b54379c1f4fa395b28/proposal/28
And my setup looks like this:
```ts
const space = '0x07e6e9047eb910f84f7e3b86cea7b1d7779c109c970a39b54379c1f4fa395b28';
const proposalId = 28;
const executor = '0x21dda40770f4317582251cffd5a0202d6b223dc167e5c8db25dc887d11eba81';

const transactions = [
  {
    to: '0xeF8305E140ac520225DAf050e2f71d5fBcC543e7',
    value: parseEther('0.01').toHexString(),
    data: '0x',
    operation: 0,
    nonce: 0
  }
];
```

**Comment by Sekhmet** [comment]
Tried your execution (in different proposal) on [this PR](https://github.com/snapshot-labs/sx-ui/pull/354) with linked sx.js:
finalize: https://testnet-2.starkscan.co/tx/0x6b2a0cc7076467ea91edf6a38062a7ac641aa146db188f2699653cf729d3767
receive: https://goerli.etherscan.io/tx/0x8a3c3fada44331cb10494c64667d0076d5eb4de55d905051ea547bb9a6f675f6
execute: https://goerli.etherscan.io/tx/0xe01412ded358860ac4b70cfbc9eca39fe4724192892cdebb6078d59e5a32d136

Can you try with new proposal using that PR and linked sx.js from this PR?

**Comment by bonustrack** [comment]
Got this error now when creating a proposal with the UI (sekhmet/sx-upgrade + yarn link to this PR)
image

**Comment by Sekhmet** [comment]
@bonustrack did you run `yarn dev --force` to reload cached deps? :thinking:

**Comment by Sekhmet** [comment]
@bonustrack if that doesn't solve maybe try pulling this branch again as I rebased it.

**Comment by bonustrack** [comment]
@Sekhmet Works now thanks!

**Comment by bonustrack** [comment]
I was able to finalize, here is the tx:
https://testnet-2.starkscan.co/tx/0x011ed0adbba843087728ed187d11445efc151ea35724246bd602b486b31c6dea
But receiveProposal and executeProposalTx failed with `cannot estimate gas` any idea what's wrong?

**Comment by Sekhmet** [comment]
> But receiveProposal and executeProposalTx failed with cannot estimate gas any idea what's wrong?

It might be because finalization transaction hasn't been received on L1 yet (it's accepted on L2 only right now). You have to wait for L1 to receive the message to progress further.

**Comment by bonustrack** [comment]
@Sekhmet Here is receive fn: https://goerli.etherscan.io/tx/0x05a7482c0bbb82318ddf097507384f832e6e8d9bb634529859e4cd26b1ec9518
But for execute I still got the error `cannot estimate gas`

**Comment by Sekhmet** [comment]
@bonustrack what proposalIndex do you use? It's not the same proposalId from Starknet, it's indexed by the number of received proposals on L1. In this case it should be `2`. I added more details about how to read it in test plan today.

**Review by bonustrack — APPROVED** [review]
tACK, works well! Here is the tx for executeProposalTx: https://goerli.etherscan.io/tx/0x315293f8ef3580937c420b9cb238e9bff60692142fb28113aa37b8d6000d580b I can see 0.01 ETH been sent to my account, feels great 😃

**Comment by Orland0x** [comment]
> tACK, works well! Here is the tx for executeProposalTx: https://goerli.etherscan.io/tx/0x315293f8ef3580937c420b9cb238e9bff60692142fb28113aa37b8d6000d580b I can see 0.01 ETH been sent to my account, feels great 😃

oi, you are stealing my eth! ;)

**Comment by bonustrack** [comment]
@Orland0x This is not a steal, it was voted by the DAO 😸

**Pull request merged**